anchored-keys/src/main/java/org/infinispan/anchored/impl/AnchoredFetchInterceptor.java

```diff
@@ -267,75 +266,42 @@ public FetchResponseCollector(K key) {
-      }
-
-      @Override
-      public CloseableIterator<CacheEntry<K, V>> iterator() {
-         // It's safe to use entrySet.iterator() because everything happens on the originator
-         return new BackingIterator(entrySet.iterator());
-      }
-
-      @Override
-      public CloseableSpliterator<CacheEntry<K, V>> spliterator() {
-         return Closeables.spliterator(iterator(), Long.MAX_VALUE,
-                                       Spliterator.CONCURRENT | Spliterator.DISTINCT | Spliterator.NONNULL);
-      }
-   }
-
-   private class BackingIterator implements CloseableIterator<CacheEntry<K, V>> {
-
-      private Iterator<CacheEntry<K, V>> iterator;
+   private class BackingEntrySet extends InternalCacheSet<CacheEntry<K, V>> {
+      protected final CacheSet<CacheEntry<K, V>> next;
 
-      public BackingIterator(Iterator<CacheEntry<K, V>> iterator) {
-         this.iterator = iterator;
+      public BackingEntrySet(CacheSet<CacheEntry<K, V>> next) {
+         this.next = next;
       }
 
       @Override
-      public void close() {
-         if (iterator instanceof CloseableIterator<?>) {
-            ((CloseableIterator<?>) iterator).close();
-         }
-      }
-
-      @Override
-      public boolean hasNext() {
-         if (iterator == null) {
-            return false;
-         }
-         return iterator.hasNext();
+      public Publisher<CacheEntry<K, V>> localPublisher(int segment) {
+         return fixPublisher(next.localPublisher(segment), segment);
       }
 
       @Override
-      public CacheEntry<K, V> next() {
-         if (iterator == null) {
-            throw new NoSuchElementException();
-         }
-         CacheEntry<K, V> localEntry = iterator.next();
-         if (localEntry.getMetadata() instanceof RemoteMetadata) {
-            // The key exists and the anchor location is a remote node
-            RemoteMetadata remoteMetadata = (RemoteMetadata) localEntry.getMetadata();
-            Address keyLocation = remoteMetadata.getAddress();
-
-            int segment = distributionManager.getCacheTopology().getSegment(localEntry.getKey());
-            return CompletionStages.join(getRemoteValue(keyLocation, localEntry.getKey(), segment, false));
-         }
-         return localEntry;
+      public Publisher<CacheEntry<K, V>> localPublisher(IntSet segments) {
+         return fixPublisher(next.localPublisher(segments), null);
       }
 
-      @Override
-      public void remove() {
-
+      private Flowable<CacheEntry<K, V>> fixPublisher(Publisher<CacheEntry<K, V>> nextPublisher, Integer segment) {
+         return Flowable.fromPublisher(nextPublisher)
+                        .groupBy(entry -> entry.getMetadata() instanceof RemoteMetadata)
+                        .flatMap(gf -> {
+                           if (!gf.getKey()) {
+                              return gf;
+                           }
+                           return gf.concatMapMaybe(incompleteLocalEntry -> {
```

Comment:
Guessing the spacing is off again for the brackets to be needed? :D

Reply:
The parameter name was too long... I renamed to `incompleteEntry` and now it fits on one line :)